### PR TITLE
0-stor tlog integration

### LINKED ISSUE #311: tlog server should use 0-stor-lib (golang)
suggestion
- remove ardb backend code for tlogserver
- use 0-stor-lib: see https://github.com/zero-os/0-stor-lib/blob/master/specs/principles.md

this has following benefits
- 1 single code path for all our storage needs
- etcd is redundant store for the metadata
- etcd can be store for double linked list

these components need to be changed to use 0-stor-lib:
- [x] flusher
- [x] slave sync (only the metadata) -> disabled in nbdserver now
- [x] tlog decoder
- [ ] manual end to end test 
- [ ] docs
- [ ] command line option

### LINKED ISSUE #374: some broken (cgo) test in nbdserver module
It must have something to do with a tlog-related test, as that's the only real thing (and certainly unstable part of that module) it is testing. This error has made travis fail since quite a while ago, but I don't seem to find a report about it, so adding it here.

Sadly the log doesn't say much about this issue:

```
58.33s$ TIMEOUT=5m make testcgo
GODEBUG=cgocheck=0 go test -timeout 5m -tags 'isal' github.com/zero-os/0-Disk github.com/zero-os/0-Disk/config github.com/zero-os/0-Disk/docs/assets github.com/zero-os/0-Disk/log github.com/zero-os/0-Disk/nbd/ardb github.com/zero-os/0-Disk/nbd/ardb/storage github.com/zero-os/0-Disk/nbd/ardb/storage/lba github.com/zero-os/0-Disk/nbd/gonbdserver github.com/zero-os/0-Disk/nbd/gonbdserver/nbd github.com/zero-os/0-Disk/nbd/nbdserver github.com/zero-os/0-Disk/redisstub github.com/zero-os/0-Disk/tlog github.com/zero-os/0-Disk/tlog/schema github.com/zero-os/0-Disk/tlog/tlogclient github.com/zero-os/0-Disk/tlog/tlogclient/blockbuffer github.com/zero-os/0-Disk/tlog/tlogclient/decoder github.com/zero-os/0-Disk/tlog/tlogclient/examples/send_tlog github.com/zero-os/0-Disk/tlog/tlogclient/examples/walk_history github.com/zero-os/0-Disk/tlog/tlogclient/player github.com/zero-os/0-Disk/tlog/tlogserver github.com/zero-os/0-Disk/tlog/tlogserver/aggmq github.com/zero-os/0-Disk/tlog/tlogserver/erasure github.com/zero-os/0-Disk/tlog/tlogserver/server github.com/zero-os/0-Disk/tlog/tlogserver/slavesync github.com/zero-os/0-Disk/tools/import_file_into_etcd_config github.com/zero-os/0-Disk/tools/key_rename_milestone_6 github.com/zero-os/0-Disk/zeroctl github.com/zero-os/0-Disk/zeroctl/cmd github.com/zero-os/0-Disk/zeroctl/cmd/config github.com/zero-os/0-Disk/zeroctl/cmd/copyvdisk github.com/zero-os/0-Disk/zeroctl/cmd/delvdisk github.com/zero-os/0-Disk/zeroctl/cmd/list github.com/zero-os/0-Disk/zeroctl/cmd/restore
ok  	github.com/zero-os/0-Disk	0.005s
ok  	github.com/zero-os/0-Disk/config	2.132s
?   	github.com/zero-os/0-Disk/docs/assets	[no test files]
?   	github.com/zero-os/0-Disk/log	[no test files]
?   	github.com/zero-os/0-Disk/nbd/ardb	[no test files]
ok  	github.com/zero-os/0-Disk/nbd/ardb/storage	4.092s
ok  	github.com/zero-os/0-Disk/nbd/ardb/storage/lba	0.018s
?   	github.com/zero-os/0-Disk/nbd/gonbdserver	[no test files]
ok  	github.com/zero-os/0-Disk/nbd/gonbdserver/nbd	23.229s
t=2017-08-06T20:41:25+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33856" module=global caller=redis.go:58
FAIL	github.com/zero-os/0-Disk/nbd/nbdserver	0.029s
?   	github.com/zero-os/0-Disk/redisstub	[no test files]
ok  	github.com/zero-os/0-Disk/tlog	0.017s
?   	github.com/zero-os/0-Disk/tlog/schema	[no test files]
ok  	github.com/zero-os/0-Disk/tlog/tlogclient	12.739s
?   	github.com/zero-os/0-Disk/tlog/tlogclient/blockbuffer	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/decoder	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/examples/send_tlog	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/examples/walk_history	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/player	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogserver	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogserver/aggmq	[no test files]
ok  	github.com/zero-os/0-Disk/tlog/tlogserver/erasure	0.003s [no tests to run]
ok  	github.com/zero-os/0-Disk/tlog/tlogserver/server	3.243s
?   	github.com/zero-os/0-Disk/tlog/tlogserver/slavesync	[no test files]
?   	github.com/zero-os/0-Disk/tools/import_file_into_etcd_config	[no test files]
?   	github.com/zero-os/0-Disk/tools/key_rename_milestone_6	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/config	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/copyvdisk	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/delvdisk	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/list	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/restore	[no test files]
make: *** [testcgo] Error 1
```

Log is from master, and a simple restart of the job fixes it most of the time, other times it requires 2 restarts.

Perhaps best not to focus on this until after 0-stor is integrated.

Questions:

+ Will 0-stor be pure go?
+ If so, I guess we can drop the CGO flags and tests? As that would remove all cgo requirements of this project...

### LINKED ISSUE #381: take a look at the vdisk_complete and improve where needed
+ Right now we can still add new vdisks once the Wait is finished;

### PR DESCRIPTION
**Pull request by iwanbk**
Fixes #311 

Notable changes:
- use 0-stor (of course)
- remove testcgo : 0-stor is pure Go
-  Fixes #381 rework vdisk completion https://github.com/zero-os/0-Disk/issues/381#issuecomment-323253812
- ~~split the test into more jobs https://github.com/zero-os/0-Disk/issues/311#issuecomment-322694804~~ cancelled by 2f6e27d37cbd6f33a8d3e706ef6b6bec5d3a45fb, 27e2d383f6e63dd64eab3aa329f3602de420a5cd
- Fixes #374 by removing the cgo test
- update docs.

### CODE REVIEW

**Comment by iwanbk** [comment]
Oh, forgot the hot reload part

**Comment by iwanbk** [comment]
> Oh, forgot the hot reload part

Implemented few hours ago.

Now investigating the issue on #404

**Comment by iwanbk** [comment]
The only thing remains is update the vendoring, as the master moved to new https://github.com/golang/dep from godep

**Comment by iwanbk** [comment]
some test become failing after merging from master

**Comment by iwanbk** [comment]
> some test become failing after merging from master

It seems to be high memory / cpu usage again, because:
- there is no error message in travis
- in my small VM, the test either `killed` or `out of memory`

From test profiling, badger sometimes use high memory usage, 588MB

**Comment by iwanbk** [comment]
I tried to reduce badger memory usage at https://github.com/zero-os/0-Disk/commit/50e3f6f486b5b7ae371d25392475eecf6d0daada

It seems to work for travis and my small VM.

Before this, my small VM always failed in race test

**Comment by iwanbk** [comment]
This PR is ready now

File: tlog/stor/walk.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+			agg, err := c.decodeCapnp(res.Data)
+			if err != nil {
+				wr.Err = err
+				wrCh <- wr
+				return
+			}
+
+			// check the epoch
+			epoch := agg.Timestamp()
+			if epoch < fromEpoch {
```
Comment (GlenDC) [inline tlog/stor/walk.go]: should this ever happen, seems like a bug to me in the `0-stor/storClient` if this would ever happen. We specifically give the fromEpoch to that API, so it seems like it should ever give something older than the given fromEpoch...

File: tlog/stor/walk.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+				// still long way to go, proceed to next key
+				continue
+			}
+			if epoch > toEpoch {
```
Comment (GlenDC) [inline tlog/stor/walk.go]: should this ever happen, seems like a bug to me in the `0-stor/storClient` if this would ever happen. We specifically give the toEpoch to that API, so it seems like it should ever give something new than the given toEpoch...

File: tlog/tlogclient/decoder/last_hash.go [hunk 3]
```diff
@@ -2,93 +2,9 @@ package decoder
...
-	"fmt"
-
-	"github.com/garyburd/redigo/redis"
-
-	"github.com/zero-os/0-Disk/tlog"
 )
 
 var (
 	// ErrNilLastHash indicates that there is no last hash entry
 	ErrNilLastHash = errors.New("nil last hash")
```
Comment (GlenDC) [inline tlog/tlogclient/decoder/last_hash.go]: this feel seems pretty useless now, either rename the file to `errors.go`, or more it to the file where it is used (as in returned), if it is only ever used in 1 file (the error that is).

File: tlog/tlogclient/decoder/limiter.go [hunk 4]
```diff
@@ -32,12 +38,23 @@ type LimitByTimestamp struct {
 
 // NewLimitByTimestamp creates new LimitByTimestamp limiter
 func NewLimitByTimestamp(startTs, endTs uint64) LimitByTimestamp {
+	if endTs == 0 {
+		endTs = uint64(time.Now().UnixNano())
```
Comment (GlenDC) [inline tlog/tlogclient/decoder/limiter.go]: Correct me if I'm wrong, but I think you better change this to `time.Now().UTC().UnixNano()`, because once you go to epoch values you lose the timezone and location information, thus you might get in weird issues if those values are used at 2 different servers. Not sure if that's the case, but in general it's safer to keep all times in the same timezone if possible. UTC is a nice standard imho, although it is as good as any I suppose.

File: tlog/tlogclient/decoder/limiter.go [hunk 5]
```diff
@@ -78,16 +95,26 @@ func (lbt LimitByTimestamp) StartBlock(block schema.TlogBlock) bool {
...
 // NewLimitBySequence creates new LimitBySequence object
 func NewLimitBySequence(startSeq, endSeq uint64) LimitBySequence {
 	return LimitBySequence{
 		startSeq: startSeq,
 		endSeq:   endSeq,
+		endTs:    uint64(time.Now().UnixNano()),
 	}
 }
 
+func (lbs LimitBySequence) FromEpoch() uint64 {
```
Comment (GlenDC) [inline tlog/tlogclient/decoder/limiter.go]: Why does a `LimitBySequence` does have a `ToEpoch`, but not a `FromEpoch` defined?

File: tlog/tlogserver/config.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+zeroStorClusters:
```
Comment (GlenDC) [inline tlog/tlogserver/config.yaml]: I suppose you uploaded this by accident, please delete it.

File: tlog/tlogserver/server/vdisk.go [hunk 7]
```diff
@@ -46,10 +46,8 @@ type vdiskFlusherCmd struct {
 type vdiskCleanupFunc func(vdiskID string)
 
 type vdisk struct {
-	id          string
-	lastHash    []byte              // current in memory last hash
-	lastHashKey []byte              // redis key of the last hash
-	respChan    chan *BlockResponse // channel of responses to be sent to client
+	id       string
+	respChan chan *BlockResponse // channel of responses to be sent to client
```
Comment (GlenDC) [inline tlog/tlogserver/server/vdisk.go]: Why give BlockResponse by reference over a channel instead of by value?

**Comment by GlenDC** [comment]
I haven't checked any of the 0-stor code, so it's not too easy for me to review this PR. Tried my best anyhow.

Also next time please squash your commits already next time, with the vendor code update as a separate commit, would make it a lot easier to review..

File: Makefile [hunk 8]
```diff
@@ -53,12 +53,12 @@ testrace: testrace_core testrace_gonbdserver
 testrace_core:
 	go test -race -timeout $(TIMEOUT) $(RACE_PACKAGES)
 
+testrace_tlog:
```
Comment (iwanbk) [inline Makefile]: done

File: Gopkg.toml [hunk 9]
```diff
@@ -4,7 +4,6 @@
   name = "github.com/asaskevich/govalidator"
 
 [[constraint]]
-  branch = "master"
   name = "github.com/coreos/etcd"
```
Comment (iwanbk) [inline Gopkg.toml]: done

File: nbd/nbdserver/tlog_test.go [hunk 10]
```diff
@@ -177,32 +182,25 @@ func testTlogStorageForceFlush(ctx context.Context, t *testing.T, vdiskID string
...
 		PrivKey:    "12345678901234567890123456789012",
-		HexNonce:   "37b8e8a308c354048d245f6d",
 	}
 
-	// create inmemory redis pool factory
-	poolFactory := tlog.InMemoryRedisPoolFactory(testConf.RequiredDataServers())
-	if !assert.NotNil(t, poolFactory) {
-		return ""
-	}
+	cleanup, configSource, _ := newZeroStorConfig(t, vdiskID, testConf)
```
Comment (iwanbk) [inline nbd/nbdserver/tlog_test.go]: fixed

File: nbd/nbdserver/tlog_test.go [hunk 11]
```diff
@@ -888,3 +887,58 @@ func TestDataHistory(t *testing.T) {
 func init() {
 	log.SetLevel(log.DebugLevel)
 }
+
+func newZeroStorConfig(t *testing.T, vdiskID string, tlogConf *server.Config) (func(), *config.StubSource, stor.Config) {
+
+	// stor server
+	storCluster, err := embeddedserver.NewZeroStorCluster(tlogConf.K + tlogConf.M)
```
Comment (iwanbk) [inline nbd/nbdserver/tlog_test.go]: Because the only thing needed by `NewZeroStorCluster` is the number of servers it need to create

File: nbd/nbdserver/tlog_test.go [hunk 12]
```diff
@@ -888,3 +887,58 @@ func TestDataHistory(t *testing.T) {
...
+		})
+	}
+
+	// meta server
+	mdServer, err := embedserver.New()
+	require.Nil(t, err)
+
+	storConf := stor.Config{
+		VdiskID:         vdiskID,
+		Organization:    os.Getenv("iyo_organization"),
```
Comment (iwanbk) [inline nbd/nbdserver/tlog_test.go]: Because 0-stor server  still need valid IYO jwt token in test, which require valid IYO credentials.
And we don't want to share our IYO credentials in our repo.

0-stor server also still need valid IYO token

File: nbd/nbdserver/vdisk_complete.go [hunk 13]
```diff
@@ -1,16 +1,30 @@
...
 type vdiskCompletion struct {
-	wg     sync.WaitGroup
-	mux    sync.Mutex
-	addMux sync.Mutex
-	errors []error
+	wg         sync.WaitGroup
+	mux        sync.Mutex
+	addMux     sync.Mutex
+	errors     []error
+	ctx        context.Context
```
Comment (iwanbk) [inline nbd/nbdserver/vdisk_complete.go]: It is used by `Stopped`

File: tlog/redispool.go [hunk 14]
```diff
@@ -165,19 +165,7 @@ func InMemoryRedisPool(requiredDataServerCount int) RedisPool {
 // using the storage cluster defined in the given Blokstor config file,
 // for that vdisk.
```
Comment (iwanbk) [inline tlog/redispool.go]: OK, also removed the crypto code

File: tlog/stor/client.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+	metaCli, err := newEtcdClient(conf.MetaShards)
+	if err != nil {
+		return nil, err
+	}
+
+	cli := &Client{
+		vdiskID:          conf.VdiskID,
+		storClient:       sc,
+		capnpBuf:         make([]byte, 0, 4096*4),
```
Comment (iwanbk) [inline tlog/stor/client.go]: yes

File: tlog/stor/client.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+		return nil, fmt.Errorf("empty meta returned by stor client")
+	}
+
+	c.lastMd = lastMd
+	c.lastMetaKey = key
+	return data, c.saveLastMetaKey()
+}
+
+func (c *Client) Close() error {
+	return nil
```
Comment (iwanbk) [inline tlog/stor/client.go]: unfortunately 0-stor client doesn't have the close method.
Yep, it must be added.
Will add it here after 0-stor client has it

Comment (iwanbk) [inline tlog/stor/client.go]: Addressed at #408

File: tlog/stor/client_test.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+	require.Nil(t, err)
+	defer mdServer.Stop()
+
+	storCluster, err := embeddedserver.NewZeroStorCluster(3)
+	require.Nil(t, err)
+	defer storCluster.Close()
+
+	conf := Config{
+		VdiskID:         vdiskID,
+		Organization:    os.Getenv("iyo_organization"),
```
Comment (iwanbk) [inline tlog/stor/client_test.go]: same comment as before

File: tlog/stor/config_source.go [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+	return Config{
+		VdiskID:         vdiskID,
+		Organization:    zsc.IYO.Org,
+		Namespace:       zsc.IYO.Namespace,
+		IyoClientID:     zsc.IYO.ClientID,
+		IyoSecret:       zsc.IYO.Secret,
+		ZeroStorShards:  serverAddrs,
+		MetaShards:      metaServerAddrs,
+		DataShardsNum:   k,
```
Comment (iwanbk) [inline tlog/stor/config_source.go]: >  why don't we use those terms in our tlog module as well, instead of the more cryptic (domain-specialized?) names k and m.

We have discussed it before in our chat :)

Yesterday someone from cairo team asked me about the meaning of this `data` and `parity` shards. I gave him this URL http://searchstorage.techtarget.com/definition/erasure-coding.
And after that he use K & M

File: tlog/stor/embeddedserver/stor.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+	server storserver.StoreServer
+}
+
+func newZeroStorServer() (*zeroStorServer, error) {
+	tmpDir, err := ioutil.TempDir("", "embedzerostor")
+	if err != nil {
+		return nil, err
+	}
+
+	opts := badgerkv.DefaultOptions
```
Comment (iwanbk) [inline tlog/stor/embeddedserver/stor.go]: Because as you might already know, 0-stor has not enough workforce to do all nice works.

More story, chris actually has started to provide in memory implementation, i think the problem is time

File: tlog/stor/meta.go [hunk 20]
```diff
@@ -0,0 +1,47 @@
+package stor
+
+import (
+	"context"
+
+	"github.com/coreos/etcd/clientv3"
+)
+
+func newEtcdClient(shards []string) (*clientv3.Client, error) {
+	return clientv3.New(clientv3.Config{
```
Comment (iwanbk) [inline tlog/stor/meta.go]: ok

File: tlog/stor/meta.go [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+		Endpoints: shards,
+	})
+}
+
+func (c *Client) getRawMeta(key []byte) ([]byte, error) {
+	resp, err := c.metaCli.Get(context.TODO(), string(key))
```
Comment (iwanbk) [inline tlog/stor/meta.go]: fixed, i use context.WithTimeout instead

File: tlog/stor/walk.go (hunk 1 above)
Comment (iwanbk) [inline tlog/stor/walk.go]: we double checking it, because 0-stor doesn't know about our data format

File: tlog/stor/walk.go (hunk 2 above)
Comment (iwanbk) [inline tlog/stor/walk.go]: we double checking it, because 0-stor doesn't know about our data format

File: tlog/tlogclient/decoder/last_hash.go (hunk 3 above)
Comment (iwanbk) [inline tlog/tlogclient/decoder/last_hash.go]: Still used by slave sync, will do with #401

File: tlog/tlogclient/decoder/limiter.go (hunk 4 above)
Comment (iwanbk) [inline tlog/tlogclient/decoder/limiter.go]: Deserves separate issue, because the ardb tlog also don't use UTC.
Created the issue : #407

File: tlog/tlogclient/decoder/limiter.go (hunk 5 above)
Comment (iwanbk) [inline tlog/tlogclient/decoder/limiter.go]: In `LimitBySequence`, the epoch is always 0 - current epoch.

File: tlog/tlogserver/config.yaml (hunk 6 above)
Comment (iwanbk) [inline tlog/tlogserver/config.yaml]: oops

File: tlog/tlogserver/server/vdisk.go (hunk 7 above)
Comment (iwanbk) [inline tlog/tlogserver/server/vdisk.go]: Not sure, it was old commit

**Comment by iwanbk** [comment]
> I haven't checked any of the 0-stor code, so it's not too easy for me to review this PR. Tried my best anyhow.

Thanks

> Also next time please squash your commits already next time, with the vendor code update as a separate commit, would make it a lot easier to review..

There might be some commits that still need to be squashed, but actually it already squashed.

Vendor code update happened 3 times:
- the initial update, in the start of work
- when merging back from master, yesterday, to resolve the conflict
- update latest 0-stor

File: nbd/nbdserver/tlog_test.go (hunk 12 above)
Comment (GlenDC) [inline nbd/nbdserver/tlog_test.go]: I think that's an issue, for unit testing that should definitely not be needed. Please make an issue for it that we don't require a valid IYO jwt token to do unit tests. That's really not a concern of our unit tests.

File: nbd/nbdserver/vdisk_complete.go (hunk 13 above)
Comment (GlenDC) [inline nbd/nbdserver/vdisk_complete.go]: Yeah sure I can see that, but that (cancel) context (created by `WithCancel`)  is never used in any other place, so I don't see how it has anything to do with stopping all vdisks...

File: tlog/stor/config_source.go (hunk 18 above)
Comment (GlenDC) [inline tlog/stor/config_source.go]: Hmm, still I think data and parity is more clear than K & M. Data and Parity contain meaning with the naming, while K & M could just as well be A & B, there is no meaning behind them besides that article (and perhaps some others) who use K & M. I don't think there is a convention for it, and even so that convention wouldn't be clear for any new reader of the code. So I would suggest to use Data & parity everywhere, ratter than K & M...

We certainly discussed this before a long time ago, but back than I only agreed because we didn't seem to have a better naming as far as I can recall. While now I see that you have the nice Data and Parity names, which I do like.

File: tlog/stor/embeddedserver/stor.go (hunk 19 above)
Comment (GlenDC) [inline tlog/stor/embeddedserver/stor.go]: Time? An in-memory implementation can be made in like less than an hour, including tests... Unless I'm missing something complex here, but it looks very simple to me.

File: tlog/stor/meta.go (hunk 21 above)
Comment (GlenDC) [inline tlog/stor/meta.go]: I don't think that's required though. The timeout can be set when creating the client with the configuration AFAIK.

File: tlog/stor/walk.go (hunk 1 above)
Comment (GlenDC) [inline tlog/stor/walk.go]: Why would they need to know about the format... We are giving the epoch from and to ranges, I really don't see why they would be different for us. Double checking is what unit tests are for. Or am I missing something here. If I miss something, please try to explain me in detail with references, as you might have the right reasons to do what you do here.

File: tlog/tlogserver/server/vdisk.go (hunk 7 above)
Comment (GlenDC) [inline tlog/tlogserver/server/vdisk.go]: I think it's safer to not pass anything by pointers over channels, as sharing pointers between goroutines can give weird results I think.

**Comment by GlenDC** [comment]
27 commits honestly doesn't look squashed to me...

File: nbd/nbdserver/tlog_test.go (hunk 12 above)
Comment (iwanbk) [inline nbd/nbdserver/tlog_test.go]: https://github.com/zero-os/0-stor/issues/182

File: nbd/nbdserver/vdisk_complete.go (hunk 13 above)
Comment (iwanbk) [inline nbd/nbdserver/vdisk_complete.go]: https://github.com/zero-os/0-Disk/blob/zerostor/nbd/nbdserver/backend.go#L231

File: tlog/stor/config_source.go (hunk 18 above)
Comment (iwanbk) [inline tlog/stor/config_source.go]: Feel free to create the issue, but won't be included in this PR because it is out of scope

File: tlog/stor/embeddedserver/stor.go (hunk 19 above)
Comment (iwanbk) [inline tlog/stor/embeddedserver/stor.go]: Never check it by myself, because i think the on disk impl is already good enough

File: tlog/stor/walk.go (hunk 1 above)
Comment (iwanbk) [inline tlog/stor/walk.go]: The 0-stor walk is based on the metadata epoch, while tlog walk is based on aggregation epoch.
There might be  < 1 second difference

Comment (GlenDC) [inline tlog/stor/walk.go]: And will this never turn into a bug, where because of this 1 second difference we might miss some aggregations which according to our data-formatted timestamp is still within range?

Comment (iwanbk) [inline tlog/stor/walk.go]: you're right, fixing it

Comment (iwanbk) [inline tlog/stor/walk.go]: fixed

Comment (iwanbk) [inline tlog/stor/walk.go]: Ups, made a mistake, wait

Comment (iwanbk) [inline tlog/stor/walk.go]: really fixed now :)

File: Gopkg.toml (hunk 9 above)
Comment (zaibon) [inline Gopkg.toml]: I also plan to move 0-stor to use dep but that's going to be for another pr

File: nbd/nbdserver/tlog_test.go (hunk 12 above)
Comment (zaibon) [inline nbd/nbdserver/tlog_test.go]: I agree if we can avoid it for unit testing it would be better. But we use the jwt for acl on namespace. If this test doesn't need to validate acl then the probability can avoid jwt here.

File: tlog/stor/config_source.go (hunk 18 above)
Comment (zaibon) [inline tlog/stor/config_source.go]: I also like more explicit name, so I vote for data and parity

File: tlog/stor/embeddedserver/stor.go (hunk 19 above)
Comment (zaibon) [inline tlog/stor/embeddedserver/stor.go]: I  have a not finish in memory implementation. It is indeed simple, but I had a bug because map doesn't keep order and this was making some test fails. 
I'll try to finish it so we can replace badger with in memory for some tests

Comment (GlenDC) [inline tlog/stor/embeddedserver/stor.go]: I don't think a KV storage would ever promise to keep all data in order anyhow, so I don't get how that could lead to a bug o.O

File: tlog/stor/walk.go (hunk 2 above)
Comment (iwanbk) [inline tlog/stor/walk.go]: need to wait for https://github.com/zero-os/0-stor/pull/184

File: tlog/stor/config_source.go (hunk 18 above)
Comment (GlenDC) [inline tlog/stor/config_source.go]: Created an issue for it: https://github.com/zero-os/0-Disk/issues/411

**Comment by GlenDC** [comment]
I approve this PR, however:

+ I haven't tested the code of this PR myself, so I'm trusting @iwanbk that this code works as intended;
+ I haven't seen the 0-Stor code, so there might be a lot of stuff to be improved, but as long as it all works as it is, that's already good enough for the time being;

**Comment by GlenDC** [comment]
Please do squash the commits before merging into master, such that we are only left with 4-5 commits or less.
